EIPS/eip-4430.md

```diff
@@ -0,0 +1,147 @@
+---
+eip: 4430
+title: Described Transactions
+description: A technique for contract authors to enable wallets to provide a human-readable description of the effect of of a transaction with a given contract.
+author: Richard Moore (@ricmoo), Nick Johnson (@arachnid)
+discussions-to: https://github.com/ethereum/EIPs/issues/4431
+status: Draft
+type: Standards Track
+category: ERC
+created: 2021-11-07
+---
+
+## Abstract
+
+Use a contract method to provide *virtual funcions* which can generate
+a human-readable description at the same time as the machine-readable
+bytecode, allowing the user to agree to the human-readable component
+in a UI while the machine can execute the bytecode once accepted.
+
+
+## Motivation
+
+When using an Ethereum Wallet (e.g. MetaMask, Clef, Hardware Wallets)
+users must accept a transaction before it can be submitted (or the user
+may decline).
+
+Due to the complexity of Ethereum transactions, wallets are very limited
+in their ability to provide insight into the effects of a transaction
+that the user is approving; outside special-cased support for common
+transactions such as ERC20 transfers, this often amounts to asking the
+user to sign an opaque blob of binary data.
+
+This EIP presents a method for dapp developers to enable a more comfortable
+user experience by providing wallets with a means to generate a better
+description about what the contract claims will happen.
+
+It does not address malicious contracts which wish to lie, it only addresses
+honest contracts that want to make their user's life better. We believe
+that this is a reasonable security model, as transaction descriptions can be
+audited at the same time as contract code, allowing auditors and code
+reviewers to check that transaction descriptions are accurate as part of
+their review.
+
+
+## Specification
+
+The **description** (a string) and the matching **execcode** (bytecode)
+are generated simultaneously by evaluating the method on a contract:
+
+```solidity
+function eipXXXDescribe(bytes inputs, bytes32 reserved) view returns (string description, bytes execcode)
```

Comment:
Also, I don't fully understand the spec. Is there a function actually named `eip4430Describe` that needs to be present in the contract in order to leverage this EIP? I suspect not since there could only be one per contract (or per argument set?) whereas the contract may contain many, if so, what is the specification for the name (and arguments) of a EIP-4430 compliant function?

Reply:
For now I want to avoid giving it a real name, since people might start using it before its ready and I want to avoid a `signTypeData_v4` situation.